R/calculate_study_day.R

```diff
@@ -0,0 +1,104 @@
+#' `calculate_study_day` performs study day calculation
+#' @description
+#' This function takes the an input data.frame and a reference data.frame (which
+#' is DM domain in most cases), and calculate the study day from reference date
+#' and target date. In case of unexpected conditions like reference date is not
+#' unique for each patient, or reference and input dates are not actual dates,
+#' NA will be returned for those records.
+#'
+#' @md
+#' @param ds_in input data.frame that contains the target date.
+#' @param ds_dm reference date.frame that contains the reference date.
+#' @param refdt reference date from `ds_dm` that will be used as reference to
+#' calculate the study day.
+#' @param tgdt target date from `ds_in` that will be used to calcualte the study
+#' day.
+#' @param study_day_var the new study day variable name in the output. For
+#' example, AESTDY for AE domain for CMSTDY for CM domain.
+#' @param merge_key character to represents the merging key between `ds_in` and
+#' `ds_dm`.
+#'
+#' @return a data.frame that takes all columns from `ds_in` and a new variable
+#' to represent the calculated study day.
+#'
+#' @examples
+#' \dontrun{
+#'   ae <- data.frame(
+#'     USUBJID = c("study123-123", "study123-124", "study123-125"),
+#'     AESTDTC = c("2012-01-01", "2012-04-14", "2012-04-14")
+#'   )
+#'   dm <- data.frame(
+#'     USUBJID = c("study123-123", "study123-124", "study123-125"),
+#'     RFSTDTC = c("2012-02-01", "2012-04-14", NA)
+#'   )
+#'   ae$AESTDTC <- as.Date(ae$AESTDTC)
+#'   dm$RFSTDTC <- as.Date(dm$RFSTDTC)
+#'   calculate_study_day(ae, dm, "RFSTDTC", "AESTDTC", "AESTDY")
+#' }
+#'
+#'
+
+calculate_study_day <- function(ds_in,
+                                ds_dm,
```

Comment:
can we rename it to dm_domain

Reply:
Add a default value dm_domain = DM

Reply:
Hi @rammprasad we want to default `dm_domain = DM` or `dm_domain = "DM"`?

Reply:
done at https://github.com/pharmaverse/sdtm.oak/pull/20/commits/10c22337db89698e272714a59afe25890b73ccfe

One thing to note is that, if DM doesn't exist, the function will throw an error 
<img width="490" alt="Screenshot 2024-01-16 at 9 07 53 PM" src="https://github.com/pharmaverse/sdtm.oak/assets/42564519/ca10555a-7349-40ce-a48e-de2764565193">

Reply:
Hi @rammprasad now since the lintr is throwing an error here, can you please confirm if I should go with `DM` or `"DM"`?

Reply:
Hi Rosemary,

I think we are going with `"DM"` and avoid non-standard evaluation or relying on variables being defined in the environment.

Reply:
Also, I think if we are following tidy design principles, then the optional parameter should be brought to the end of the function prototype, i.e. `refdt` and `merge_key` would be the parameters at the last positions.

Reply:
Hi @ramiromagno I was told to default dm_domain to DM, so I will use "DM" and move them to the last positions then! Thanks!

Reply:
Do not use any default as discussed in the meeting